src/ringct/bulletproofs.cc

```diff
@@ -0,0 +1,750 @@
+     
+  // PAPER LINES 43-45
+  rct::keyV hashed;
+  hashed.push_back(A);
+  hashed.push_back(S);
+  rct::key y = rct::hash_to_scalar(hashed);
+  rct::key z = rct::hash_to_scalar(y);
+     
+  // Polynomial construction before PAPER LINE 46
+  rct::key t0 = rct::zero();
+  rct::key t1 = rct::zero();
+  rct::key t2 = rct::zero();
+           
+  const auto yN = vector_powers(y, N);
+
+  rct::key ip1y = inner_product(oneN, yN);
+  rct::key tmp;
+  sc_muladd(t0.bytes, z.bytes, ip1y.bytes, t0.bytes);
+
+  rct::key zsq;
+  sc_mul(zsq.bytes, z.bytes, z.bytes);
+  sc_muladd(t0.bytes, zsq.bytes, sv.bytes, t0.bytes);
+
+  rct::key k = rct::zero();
+  sc_mulsub(k.bytes, zsq.bytes, ip1y.bytes, k.bytes);
+
+  rct::key zcu;
+  sc_mul(zcu.bytes, zsq.bytes, z.bytes);
+  sc_mulsub(k.bytes, zcu.bytes, ip12.bytes, k.bytes);
+  sc_add(t0.bytes, t0.bytes, k.bytes);
+
+  // DEBUG: Test the value of t0 has the correct form
+#ifdef DEBUG_BP
+  rct::key test_t0 = rct::zero();
+  rct::key iph = inner_product(aL, hadamard(aR, yN));
+  sc_add(test_t0.bytes, test_t0.bytes, iph.bytes);
+  rct::key ips = inner_product(vector_subtract(aL, aR), yN);
+  sc_muladd(test_t0.bytes, z.bytes, ips.bytes, test_t0.bytes);
+  rct::key ipt = inner_product(twoN, aL);
+  sc_muladd(test_t0.bytes, zsq.bytes, ipt.bytes, test_t0.bytes);
+  sc_add(test_t0.bytes, test_t0.bytes, k.bytes);
+  CHECK_AND_ASSERT_THROW_MES(t0 == test_t0, "t0 check failed");
+#endif
+  PERF_TIMER_STOP(PROVE_step1);
+     
+  PERF_TIMER_START_BP(PROVE_step2);
+  const auto HyNsR = hadamard(yN, sR);
+  const auto vpIz = vector_scalar(oneN, z);
+
+  rct::key ip1 = inner_product(vector_subtract(aL, vpIz), HyNsR);
+  sc_add(t1.bytes, t1.bytes, ip1.bytes);
+
+  rct::key ip2 = inner_product(sL, vector_add(hadamard(yN, vector_add(aR, vpIz)), vector_scalar(twoN, zsq)));
+  sc_add(t1.bytes, t1.bytes, ip2.bytes);
+
+  rct::key ip3 = inner_product(sL, HyNsR);
+  sc_add(t2.bytes, t2.bytes, ip3.bytes);
+     
+  // PAPER LINES 47-48
+  rct::key tau1 = rct::skGen(), tau2 = rct::skGen();
+
+  rct::key T1 = rct::addKeys(rct::scalarmultKey(rct::H, t1), rct::scalarmultBase(tau1));
+  rct::key T2 = rct::addKeys(rct::scalarmultKey(rct::H, t2), rct::scalarmultBase(tau2));
+     
+  // PAPER LINES 49-51
+  hashed.clear();
+  hashed.push_back(z);
+  hashed.push_back(T1);
+  hashed.push_back(T2);
+  rct::key x = rct::hash_to_scalar(hashed);
+           
+  // PAPER LINES 52-53
+  rct::key taux = rct::zero();
+  sc_mul(taux.bytes, tau1.bytes, x.bytes);
+  rct::key xsq;
+  sc_mul(xsq.bytes, x.bytes, x.bytes);
+  sc_muladd(taux.bytes, tau2.bytes, xsq.bytes, taux.bytes);
+  sc_muladd(taux.bytes, gamma.bytes, zsq.bytes, taux.bytes);
+  rct::key mu;
+  sc_muladd(mu.bytes, x.bytes, rho.bytes, alpha.bytes);
+     
+  // PAPER LINES 54-57
+  rct::keyV l = vector_add(vector_subtract(aL, vpIz), vector_scalar(sL, x));
+  rct::keyV r = vector_add(hadamard(yN, vector_add(aR, vector_add(vpIz, vector_scalar(sR, x)))), vector_scalar(twoN, zsq));
```

Comment:
I can see quite some opportunities for reusing already computed data such as:
```c++
vector_subtract(aL, vpIz)
vector_add(aR, vpIz)
vector_scalar(twoN, zsq)
```

Reply:
Ah, I did not make another pass after the log modifications :/ Will fix.